rfd/0164-scoped-rbac.md

```diff
@@ -0,0 +1,956 @@
+* At a `Step 1`, at the timestamp `T1` Alice removed Bob from access list granting role `access` in the scope `/dev/lab`. Teleport has updated the `_root` nodes of grant hierarchy `/dev` and `/dev/lab` with this timestamp in the backend.
+* At a `Step 2`, at the timestamp `T2` > `T1`, Alice added a new resource `luna` to the scope of the `/dev/lab`. Teleport has updated the `_root` node of the resource group hierarchy `/dev` and `/dev/lab` with this timestamp in the backend.
+* At a `Step 3`, Teleport proxy would have to evaluate access to host luna. The proxy would have to make sure that it's copy of the grant hierarchies `/dev/lab` is not stale.
+
+In our example `T2 > T1 > Tr` and `T2 > T1 > Tg`. There are two sub cases when `Tr > Tg` and `Tr < Tg`
+
+Here are possible combinations of versions that The Proxy can encounter:
+
+```
+1. G - Tg, R - Tr
+2. G - T1, R - Tr
+3. G - Tg, R - T2
+4. G - T1, R - T2
+```
+
+Let's review the cases in the reverse order:
+
+* In the cases 3 and 4, the Proxy sees that the resource group has been updated after the grant, because `T2 > T1 > Tg` and can proceed.
+* In the case 2, The proxy sees that `R` may be stale, because it's `Tr` less than `T1 and has to make a call to the backend making sure it fetched update of `R` least as fresh as `T1` with membership of `luna`.
+* In case 1, if `Tr > Tg`, the proxy can proceed, otherwise if `Tr < Tg`, the proxy has to make sure it has fetched the latest update of `Tr` at least as fresh as `Tg` with membership of `luna`.
+
+How would Teleport "make sure that it has fetched a version of a resource at least as fresh as a timestamp?". 
+
+Some backends make it easy, like Etcd giving a generation version for every resource, making sure you are not looking at the stale snapshot. 
+
+Backends like DynamoDB are making it very hard or impossible, the only way with those backends is for the client to keep track when was the last time if has last fetched the resource `R` in cases 1 and 2, and update this value for `R` in it's in memory-cache.
+
+We also are making assumption that time is in sync between all clients. In real-world, it's more complicated, because we have to consider the times when roles have been updated and fetched, but the algorithm will remain the same.
+
+TODO: question for a scale team on implementation details for backends.
+
+### Migration
+
+We will release a new versions of resources in multiple phases:
+
+* `RoleV8` will introduce `parent_resource_group` and `grantable_scope`, `create_time`, `update_time`. `RoleV9` will remove `node_labels`,  `node_labels_expression`. Users can gradually migrate to new versions, first by leveraging scopes and resoruce groups, and then removing labels matching. For new `RoleV8` roles by default, if the `grantable_scope` is missing, we assume empty scope - that will prevent the role from being granted on any scopes. When migrating existing roles from `V7`, we would set `/` - root scope to avoid breaking the cluster. 
```

Comment:
If we get rid of e.g. app_labels we need another way to say that the role grants access to apps

Before seeing this RFD I have been considering the idea of having an `access` verb like the other RBAC verbs

Reply:
I think `access` verb is a nice idea, I added a note in the section "The access verb", take a look.